netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,305 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.transport.netty;
+
+import static io.netty.handler.codec.http2.Http2Error.PROTOCOL_ERROR;
+import static io.netty.handler.codec.http2.Http2Exception.connectionError;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder;
+import io.netty.handler.codec.http2.Http2CodecUtil;
+import io.netty.handler.codec.http2.Http2ConnectionAdapter;
+import io.netty.handler.codec.http2.Http2ConnectionEncoder;
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.handler.codec.http2.Http2Stream;
+import io.netty.util.ReferenceCountUtil;
+
+import java.util.ArrayDeque;
+import java.util.Map;
+import java.util.Queue;
+import java.util.TreeMap;
+
+/**
+ * <p>
+ * Implementation of a {@link Http2ConnectionEncoder} that dispatches all method call to
+ * another {@link Http2ConnectionEncoder}, except for when the maximum number of
+ * concurrent streams limit is reached.
+ * </p>
+ * <p>
+ * When this limit is hit, instead of rejecting any new streams this implementation
+ * buffers newly created streams and their corresponding frames. Once an active stream
+ * gets closed or the maximum number of concurrent streams is increased, this encoder will
+ * automatically try to empty its buffer and create as many new streams as possible.
+ * </p>
+ * <p>
+ * This implementation makes the buffering mostly transparent and can be used as a drop
+ * in replacement for {@link DefaultHttp2ConnectionEncoder}.
+ * </p>
+ */
+class BufferingHttp2ConnectionEncoder extends DecoratingHttp2ConnectionEncoder {
+
+  // Buffer for any streams and corresponding frames that could not be created
+  // due to the maximum concurrent stream limit being hit.
+  private final TreeMap<Integer, PendingStream> pendingStreams =
+          new TreeMap<Integer, PendingStream>();
+  // Smallest stream id whose corresponding frames do not get buffered.
+  private int largestCreatedStreamId;
+
+  static class Builder extends DecoratingHttp2ConnectionEncoder.Builder {
+
+    Builder(Http2ConnectionEncoder.Builder delegateBuilder) {
+      super(delegateBuilder);
+    }
+
+    @Override
+    protected Http2ConnectionEncoder buildInternal(Http2ConnectionEncoder delegate) {
+      return new BufferingHttp2ConnectionEncoder(delegate);
+    }
+  }
+
+  protected BufferingHttp2ConnectionEncoder(Http2ConnectionEncoder delegate) {
+    super(delegate);
+    connection().addListener(new Http2ConnectionAdapter() {
+      @Override
+      public void goingAway() {
+        failPendingStreams();
+      }
+
+      @Override
+      public void streamInactive(Http2Stream stream) {
+        createNextPendingStream();
+      }
+    });
+  }
+
+  static Builder newBuilder() {
```

Comment:
Could the static and non-static methods/fields be less shuffled? I can deal with some, but this goes non-static, static, non-static, static, non-static. Putting newBuilder() next to Builder would suffice for me.